Algo.h

```diff
@@ -129,32 +159,93 @@ class MMP
         // ps1.x()));
         double x = 0.0;
         // (x, 0) is equidistant from ps1 and ps2
+        //calculated x directly from research paper
+        double alpha = ps2.x() - ps1.x();
+        double beta = i2.ps_d - i1.ps_d;
+        double gamma = ps1.squaredLength() - ps2.squaredLength() - beta*beta;
+        //ax^2 + bx + c = 0
+        double a = alpha*alpha - beta*beta;
+        //γα + 2s1xβ2
+        double b = gamma*alpha + 2.0*ps2.x()*beta*beta;
+        //c = 1/4γ2 − s12β2
+        double c = 0.25*gamma*gamma - ps2.squaredLength()*beta*beta;
+
+        if(b*b - 4*a*c > 0)
+        {
+            double sq_det = sqrt(b*b - 4*a*c); 
+            x = (-b + sq_det)/(2*a);
+        }
+        else
+            x = -1.0;
 
         if (x >= st and x <= end) {
             // [st, x] closer to ps1 or ps2?
             if ((st_v - ps1).squaredLength() < (st_v - ps2).squaredLength()) {
-                b_intervals.push_back(Interval(ps1, st, x, i1.ps_d, i1.from, i1.edge));
-                b_intervals.push_back(Interval(ps2, x, end, i2.ps_d, i2.from, i2.edge));
+                b_intervals.push_back(Interval(ps1, st, x, i1.ps_d, i1.from, i1.edge, false));
+                b_intervals.push_back(Interval(ps2, x, end, i2.ps_d, i2.from, i2.edge, false));
             } else {
-                b_intervals.push_back(Interval(ps2, st, x, i2.ps_d, i2.from, i2.edge));
-                b_intervals.push_back(Interval(ps1, x, end, i1.ps_d, i1.from, i1.edge));
+                b_intervals.push_back(Interval(ps2, st, x, i2.ps_d, i2.from, i2.edge, false));
+                b_intervals.push_back(Interval(ps1, x, end, i1.ps_d, i1.from, i1.edge, false));
             }
         } else {
             // the complete interval is near to one single source
             if ((ps2 - st_v).squaredLength() < (ps1 - st_v).squaredLength())
-                b_intervals.push_back(Interval(ps2, st, end, i2.ps_d, i2.from, i2.edge));
+                b_intervals.push_back(Interval(ps2, st, end, i2.ps_d, i2.from, i2.edge, false));
             else
-                b_intervals.push_back(Interval(ps1, st, end, i1.ps_d, i1.from, i1.edge));
+                b_intervals.push_back(Interval(ps1, st, end, i1.ps_d, i1.from, i1.edge, false));
```

Comment:
@ahemanthkumar agreed, we do not need to invert here ever, because the intervals themselves are guaranteed to be following the precondition. But niche, we must check.